dialogue-core/src/main/java/com/palantir/dialogue/core/RttSampler.java

```diff
@@ -0,0 +1,280 @@
+            @Override
+            public void onSuccess(List<Long> result) {
+                maybePermit.get().close();
+
+                if (log.isDebugEnabled()) {
+                    List<Long> millis =
+                            result.stream().map(TimeUnit.NANOSECONDS::toMillis).collect(Collectors.toList());
+                    long[] best = Arrays.stream(rtts)
+                            .mapToLong(rtt -> rtt.getRttNanos().orElse(Long.MAX_VALUE))
+                            .toArray();
+                    log.debug(
+                            "RTTs {} {} {}",
+                            SafeArg.of("nanos", result),
+                            SafeArg.of("millis", millis),
+                            SafeArg.of("best", Arrays.toString(best)));
+                }
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                maybePermit.get().close();
+                log.info("Failed to sample RTT for channels", throwable);
+            }
+        });
+    }
+
+    @VisibleForTesting
+    enum RttEndpoint implements Endpoint {
+        INSTANCE;
+
+        @Override
+        public void renderPath(Map<String, String> _params, UrlBuilder _url) {}
+
+        @Override
+        public HttpMethod httpMethod() {
+            return HttpMethod.OPTIONS;
+        }
+
+        @Override
+        public String serviceName() {
+            return "RttSampler";
+        }
+
+        @Override
+        public String endpointName() {
+            return "rtt";
+        }
+
+        @Override
+        public String version() {
+            return UserAgentEndpointChannel.dialogueVersion();
+        }
+    }
+
+    /**
+     * Always returns the *minimum* value from the last few samples, so that we exclude slow calls that might include
+     * TLS handshakes.
+     */
+    @VisibleForTesting
+    @ThreadSafe
+    static final class RttMeasurement {
+        private static final int NUM_MEASUREMENTS = 5;
+
+        private final long[] samples;
+        private volatile long bestRttNanos = Long.MAX_VALUE;
+
+        RttMeasurement() {
+            samples = new long[NUM_MEASUREMENTS];
+            Arrays.fill(samples, Long.MAX_VALUE);
+        }
+
+        public OptionalLong getRttNanos() {
+            return bestRttNanos == Long.MAX_VALUE ? OptionalLong.empty() : OptionalLong.of(bestRttNanos);
+        }
+
+        synchronized void addMeasurement(long newMeasurement) {
+            Preconditions.checkArgument(newMeasurement > 0, "Must be greater than zero");
+            Preconditions.checkArgument(newMeasurement < Long.MAX_VALUE, "Must be less than MAX_VALUE");
+
+            if (samples[0] == Long.MAX_VALUE) {
+                Arrays.fill(samples, newMeasurement);
+                bestRttNanos = newMeasurement;
+            } else {
+                System.arraycopy(samples, 1, samples, 0, NUM_MEASUREMENTS - 1);
```

Comment:
instead of copying the array could we treat the array as a circular buffer keeping  track of our current location

Reply:
Yeah I considered this, but figured it was actually gonna be more readable to try and keep it simple here - especially when this codepath only gets invoked at most once per second, and in a non-blocking way :)